### ISSUE TITLE
`pretty_print` raises errors when asked to print any value other than an array or object

### ISSUE THREAD

**Issue by cemerick** [ocaml-community/yojson#121]
e.g.:

```
utop # Format.printf "%a@." (Yojson.Basic.pretty_print ~std:true) (`Int 5);;
Exception:
Yojson.Json_error
 "Root is not an object or array as requested by the JSON standard"
```
My first question is, where does the standard request that JSON values aside from objects and arrays not be printed? Looking at [json.org](http://json.org), and the actual [ECMA standard](https://www.ecma-international.org/wp-content/uploads/ECMA-404_2nd_edition_december_2017.pdf), I don't see any language to this effect. On the contrary, the ECMA document states plainly that "A JSON value can be an object, array, number, string, true, false, or null", and the quasi-BNF on the side of json.org is pretty clear that any of these things is a valid JSON value.

Beyond litigating the specification though, the current behaviour is just incredibly unhelpful. Javascript runtimes happily `parse` and `stringify` non-object and non-array "root" values, and OCaml programs need to deal with the same. To that end, I consistently carry around these functions so I can serialize JSON values as required by APIs and other programs (with the understanding that my treatment of strings via Format's `%S` might actually be slightly off):

```ocaml
let pp_json f fmt =
  let open Format in
  function
  | `Null -> pp_print_string fmt "null"
  | `Int v -> pp_print_int fmt v
  | `Float v -> pp_print_float fmt v
  | `String v -> fprintf fmt "%S" v
  | `Bool v -> pp_print_bool fmt v
  | v -> fprintf fmt "@[%a@]" f v

let pp_json_basic = pp_json (Yojson.Basic.pretty_print ~std:true)

let pp_json_safe = pp_json (Yojson.Safe.pretty_print ~std:true)
```

Can we remove this limitation?

**Comment by cemerick** [comment]
As an addendum re: specification litigation, [RFC8259](https://datatracker.ietf.org/doc/html/rfc8259) also talks about JSON generators producing "JSON texts", examples of which it cites include atomic values like `"Hello world!"`, `42`, and `true`. :shrug:

**Comment by Leonidas-from-XIV** [comment]
I agree with you there and if fact I once ran into an issue back in the day in AngularJS where my REST API returned `"abc"` as a valid JSON value and it got parsed as `{"a": 0, "b": 1, "c" : 2}` which is essentially parody of my response. I then also looked up into the, uhm, "standards" and found no wording that would prevent `"abc"` as valid JSON value.

Do you have time to submit this as PR?

**Comment by cemerick** [comment]
@Leonidas-from-XIV For sure, sometime in the next day or two. :+1:

**Comment by Leonidas-from-XIV** [comment]
One could argue that removing the `std` parameter (since it is false anyway) would fix that particular problem without really introducing other issues (given it only controls the non-standard Tuple/Variant extensions, that I doubt anybody would use) but I think it is also extremely silly to argue for an unhelpful/broken behavior by quoting a spec where it doesn't actually specify that behavior. So we definitely should axe this.

**Comment by cemerick** [comment]
I had actually forgotten that `std` was involved in controlling whether the "root" value had to be an array or object; I had internalized its role re: variants and `NaN`, etc., but nothing else.

@Leonidas-from-XIV #104 is still pending (?), so it feels like the PR I open immediately should leave `std` in place?

**Comment by Leonidas-from-XIV** [comment]
Ah, `NaN`-handling might be good reason to keep `std:true`, indeed. I only learned about the difference of the root handling when I looked up in the source code, where there isn't really a reason why it would be doing it this way.

For now I would leave `std` in place, given that just removing the odd root behaviour should be more or less a compatible change. Not throwing an error where there used to be one is probably fine. If we go around removing the non-standard extensions that's better for a Yojson 3.0 release and then it would make a lot of sense to drop `std` as well.

**Comment by mjambon** [comment]
It's the JSON spec from 2006 (https://datatracker.ietf.org/doc/html/rfc4627) that requires a "JSON text" be either an object or an array. [Wikipedia](https://en.wikipedia.org/wiki/JSON) summarizes the change as
> Early versions of JSON (such as specified by RFC 4627) required that a valid JSON text must consist of only an object or an array type, which could contain other types within them. This restriction was dropped in RFC 7158, where a JSON text was redefined as any serialized value.

The reasons why I obeyed the restriction at the time aren't very clear, though. I could totally have chosen to disregard it for the sake of convenience, just like I decided to skip UTF-8 validation (separation of concerns, performance) or added experimental constructs for variants and tuples.

**Comment by Leonidas-from-XIV** [comment]
@mjambon That explains a lot why it was the case, thanks! I'm thinking whether it would not be useful to add this into the documentation somewhere.

Overall it is very good that RFC 7158 dropped this requirement, because now a Yojson value is always valid JSON, whereas with the old definition it would only be valid if the top-level was a `` `List`` or `` `Assoc`` which is way worse to compose JSON values.
